cluster-api/machines-api/proposal.md

```diff
@@ -0,0 +1,164 @@
+with a new one matching the updated spec. If a Machine object is deleted, the
+corresponding Node should have its external resources released by the
+provider-specific controller, and should be deleted as well.
+
+Fields like the kubelet version, the container runtime to use, and its version,
+are modeled as fields on the Machine's spec. Any other information that is
+provider-specific, though, is part of an opaque ProviderConfig string that is
+not portable between different providers.
+
+The ProviderConfig is recommended to be a serialized API object in a format
+owned by that provider, akin to the [Component Config](https://goo.gl/opSc2o)
+pattern. This will allow the configuration to be strongly typed, versioned, and
+have as much nested depth as appropriate. These provider-specific API
+definitions are meant to live outside of the Machines API, which will allow them
+to evolve independently of it. Attributes like instance type, which network to
+use, and the OS image all belong in the ProviderConfig.
+
+## In-place vs. Replace
+
+One simplification that might be controversial in this proposal is the lack of
+API control over "in-place" versus "replace" reconciliation strategies. For
+instance, if a Machine's spec is updated with a different version of kubelet
+than is actually running, it is up to the provider-specific controller whether
+the request would best be fulfilled by performing an in-place upgrade on the
+Node, or by deleting the Node and creating a new one in its place (or reporting
+an error if this particular update is not supported). One can force a Node
+replacement by deleting and recreating the Machine object rather than updating
+it, but no similar mechanism exists to force an in-place change.
+
+Another approach considered was that modifying an existing Machine should only
+ever attempt an in-place modification to the Node, and Node replacement should
+only occur by deleting and creating a new Machine. In that case, a provider
+would set an error field in the status if it wasn't able to fulfill the
+requested in-place change (such as changing the OS image or instance type in a
+cloud provider).
+
+The reason this approach wasn't used was because most cluster upgrade tools
+built on top of the Machines API would follow the same pattern:
+
+    for machine in machines:
+        attempt to upgrade machine in-place
+        if error:
+            create new machine
+            delete old machine
+
+Since updating a Node in-place is likely going to be faster than completely
+replacing it, most tools would opt to use this pattern to attempt an in-place
+modification first, before falling back to a full replacement.
+
+It seems like a much more powerful concept to allow every tool to instead say:
+
+    for machine in machines:
+        update machine
+
+and allow the provider to decide if it is capable of performing an in-place
+update, or if a full Node replacement is necessary.
```

Comment:
Agree - and this could also be an option / policy that can be set by the machine controller.  For example, kops encourages full node replacement (a little more reliable, perhaps).  But certainly it is slower, and some people would likely choose in-place replacement if it was available.  And if we have a kops-controller, it'll _have_ to support in-place for bare-metal.